core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -659,18 +657,20 @@ public Node start() throws NodeValidationException {
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.getTaskManager().setTaskResultsService(injector.getInstance(TaskResultsService.class));
         transportService.start();
+        assert localNodeFactory.getNode() != null;
+        assert transportService.getLocalNode().equals(localNodeFactory.getNode())
+            : "transportService has a different local node than the factory provided";
         validateNodeBeforeAcceptingRequests(settings, transportService.boundAddress(), pluginsService.filterPlugins(Plugin.class).stream()
             .flatMap(p -> p.getBootstrapChecks().stream()).collect(Collectors.toList()));
 
         clusterService.addStateApplier(transportService.getTaskManager());
+        clusterService.getMasterService().setClusterStateSupplier(discovery::clusterState);
```

Comment:
Now that we separated MasterService and ClusterApplier, wouldn't it make sense to make MasterService the custodian of the last committed cluster state instead of keeping it in Discovery. I feel like it might simplify a few things in wiring dependencies here.